### PR TITLE
Add doc about the BOSHDeployment controller

### PR DESCRIPTION
**Pull request by aduffeck**

### CODE REVIEW

File: docs/controllers/boshdeployment.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+## Reconcilers
+
+### Deployment Reconciler
+
+Watches for:
+
+- `BOSHDeployment`
+- `ConfigMap`/`Secret` for ops files and the deployment manifest
+- `Pods` created for the BOSHDeployment
```
Comment (manno) [inline docs/controllers/boshdeployment.md]: https://github.com/cloudfoundry-incubator/cf-operator/blob/master/pkg/kube/controllers/boshdeployment/deployment_controller.go#L37-L44

But why?

Comment (edwardstudy) [reply docs/controllers/boshdeployment.md]: Maybe we could clean up this Pod watch if it's OK. I didn't find any Pods belonged to a boshDeployment

File: docs/controllers/boshdeployment.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+> The ["Data Gathering"](https://github.com/cloudfoundry-incubator/cf-operator/tree/master/docs/commands/cf-operator_util_data-gather.md) step generates 2 secrets for each Instance Group:
+>
+> - Instance Group Resolved Properties
+> - Instance Group BPM
+>
+> The "Versioned" `Secret` for Instance Group Resolved Properties is referenced by the Instance Group `ExtendedStatefulSets` and `ExtendedJobs`
+
+Ownership:
+
+TODO document if this will have to keep ownership on something?
```
Comment (edwardstudy) [inline docs/controllers/boshdeployment.md]: Ownership can be used in cleanup phase, so it is needed. AFAIK

File: docs/controllers/boshdeployment.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# BOSHDeployment
+
+- [BOSHDeployment](#bosheployment)
```
Comment (viovanov) [inline docs/controllers/boshdeployment.md]: typo in the link

File: docs/controllers/boshdeployment.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  - [Reconcilers](#reconcilers)
+    - [Deployment Reconciler](#deployment-reconciler)
+    - [Generated Variable Reconciler](#generated-variable-reconciler)
+    - [BPM Reconciler](#bpm-reconciler)
+
+## Description
+
+A BOSH deployment is created from a deployment manifest and optionally ops files.
+
+The deployment manifest is based on a vanilla BOSH deployment manifest,
```
Comment (viovanov) [inline docs/controllers/boshdeployment.md]: Comma at the end of the sentence

File: docs/controllers/boshdeployment.md (hunk 1 above)
Comment (viovanov) [reply docs/controllers/boshdeployment.md]: There's no reason to watch for pods at this level. If we had to, we'd be doing something wrong.

File: docs/controllers/boshdeployment.md (hunk 2 above)
Comment (viovanov) [reply docs/controllers/boshdeployment.md]: Based on docs, ownership is used for garbage collection. As a general rule, all resources we _create_ to reconcile a specific object should be owned by that object.

Do we use it for anything else, and if garbage collection is automatic, why do we still need finalizers?

Comment (manno) [reply docs/controllers/boshdeployment.md]: Finalizers are mostly used to remove ownership from objects we don't want to clean up: custom resources created by the user for example.

So when you register ownership on a 'foreign' pod, you also set a finalizer on the new owner. When deleting the owner you can remove finalizer from the owner and the ownership from the pod.

File: docs/controllers/boshdeployment.md (hunk 1 above)
Comment (manno) [reply docs/controllers/boshdeployment.md]: anybody working on removing it? Or should I create a story?

File: docs/controllers/boshdeployment.md (hunk 2 above)
Comment (edwardstudy) [reply docs/controllers/boshdeployment.md]: I thought we use ownership to make it happens that when you delete an instance, the instance's dependents(sts, jobs, services, etc) are also deleted. And `finalizers` only controls this process.

1. Without `finalizers`, this deletion is automated
2. Without ownership, this deletion doesn't happen and kube leaves orphans to us

**Pull request merged**